modules/mod_ginger_base/controllers/controller_search.erl

```diff
@@ -16,36 +16,89 @@ init(Args) ->
 content_types_provided(Req, State) ->
     {[{"application/json", to_json}], Req, State}.
 
+%% @doc Retrieve a value from nested maps
+-spec get_in([term()], map()) -> term().
+get_in([Key], Map) ->
+    maps:get(Key, Map);
+get_in([Key|Keys], Map) ->
+    get_in(Keys, maps:get(Key, Map)).
+
 to_json(Req, State) ->
     %% Init
     Context  = z_context:new(Req, ?MODULE),
     RequestArgs = wrq:req_qs(Req),
+    
     %% Get search params from request
     Type = list_to_atom(proplists:get_value("type", RequestArgs, "ginger_search")),
     Offset = list_to_integer(proplists:get_value("offset", RequestArgs, "0")),
     Limit = list_to_integer(proplists:get_value("limit", RequestArgs, "1000")),
+    
     %% Perform search (Zotonic offsets start at 1)
-    Result = z_search:search({Type, arguments(RequestArgs)}, {Offset + 1, Limit}, Context),
-    #search_result{
-        result = Results,
-        facets = _Facets,
-        total = Total
-    } = Result,
-    %% Filter search results not visible for current user
-    VisibleResults = lists:filter(
-        fun(R) ->
-            is_visible(R, Context)
-        end,
-        Results
-    ),
-    %% Serialize to JSON
-    SearchResults = #{
-        <<"result">> => [search_result(R, Context) || R <- VisibleResults],
-        <<"total">> => Total
-    },
-    Json = jsx:encode(SearchResults),
-    %% Done
-    {Json, Req, State}.
+    case proplists:get_value(mode, State) of 
+        coordinates ->
+            %% We're only interested in the geolocation and id fields
+	    Query1 = [{source, [<<"geolocation">>]} | arguments(RequestArgs)],
+            %% And it doesn't make sense to retrieve any results without coordinates
+	    Query2 = [{has_geo, <<"true">>} | Query1],
+	    SearchResults = z_search:search({Type, Query2}, {Offset + 1, Limit}, Context),
+            %% Strip any unneeded data
+	    Coordinates = 
+                lists:map(
+                  fun(Item) -> 
+                          Id = maps:get(<<"_id">>, Item),
+                          #{<<"lat">> := Lat,
+                            <<"lon">> := Lon} = 
+                              get_in([<<"_source">>, <<"geolocation">>], Item),
```

Comment:
This works, but what’s wrong with a pattern match in `fun(Item)` instead? To me, it would be clearer.